src/software/ai/hl/stp/play/shoot_or_pass_play.cpp

```diff
@@ -106,101 +146,73 @@ PassWithRating ShootOrPassPlay::attemptToShootWhileLookingForAPass(
     TacticCoroutine::push_type &yield, std::shared_ptr<GoalieTactic> goalie_tactic,
     std::array<std::shared_ptr<CreaseDefenderTactic>, 2> crease_defender_tactics,
     std::shared_ptr<ShootGoalTactic> shoot_tactic, const World &world)
+
 {
-    // If the passing is coming from the friendly end, we split the cherry-pickers
-    // across the x-axis in the enemy half
-    Rectangle cherry_pick_1_target_region = world.field().enemyPositiveYQuadrant();
-    Rectangle cherry_pick_2_target_region = world.field().enemyNegativeYQuadrant();
-
-    // Otherwise, the pass is coming from the enemy end, put the two cherry-pickers
-    // on the opposite side of the x-axis to wherever the pass is coming from
-    if (world.ball().position().x() > -1)
-    {
-        double y_offset =
-            -std::copysign(world.field().yLength() / 2, world.ball().position().y());
-        cherry_pick_1_target_region =
-            Rectangle(Point(0, 0), Point(world.field().xLength() / 4, y_offset));
-        cherry_pick_2_target_region =
-            Rectangle(Point(world.field().xLength() / 4, 0),
-                      Point(world.field().xLength() / 2, y_offset));
-    }
+    auto pitch_division =
+        std::make_shared<const EighteenZonePitchDivision>(world.field());
 
-    std::array<std::shared_ptr<CherryPickTactic>, 2> cherry_pick_tactics = {
-        std::make_shared<CherryPickTactic>(world, cherry_pick_1_target_region,
-                                           play_config->getPassingConfig()),
-        std::make_shared<CherryPickTactic>(world, cherry_pick_2_target_region,
-                                           play_config->getPassingConfig())};
+    PassGenerator<EighteenZoneId> pass_generator(pitch_division,
+                                                 play_config->getPassingConfig());
 
-    // Start a PassGenerator that will continuously optimize passes into the enemy half
-    // of the field
-    PassGenerator pass_generator(world, world.ball().position(),
-                                 PassType::RECEIVE_AND_DRIBBLE,
-                                 play_config->getPassingConfig());
-    pass_generator.setTargetRegion(world.field().enemyHalf());
-    PassWithRating best_pass_and_score_so_far = pass_generator.getBestPassSoFar();
+    auto pass_eval             = pass_generator.generatePassEvaluation(world);
+    Zones cherry_pick_region_1 = {
+        pass_eval.getBestZonesToCherryPick(world.field(), world.ball().position())[0]};
+    Zones cherry_pick_region_2 = {
+        pass_eval.getBestZonesToCherryPick(world.field(), world.ball().position())[1]};
```

Comment:
similar comments here for making sure the array values exist and avoiding duplicate function calls

Reply:
added